src/CommunityToolkit.Maui/PopupService.cs

```diff
@@ -0,0 +1,171 @@
+﻿using CommunityToolkit.Maui.Core;
+using CommunityToolkit.Maui.Views;
+using Microsoft.Maui.Controls.Platform;
+using System.ComponentModel;
+
+namespace CommunityToolkit.Maui;
+
+/// <inheritdoc cref="IPopupService"/>
+public class PopupService : IPopupService
+{
+	readonly IServiceProvider serviceProvider;
+
+	static readonly IDictionary<Type, Type> viewModelToViewMappings = new Dictionary<Type, Type>();
+
+	static Page CurrentPage =>
+		PageExtensions.GetCurrentPage(
+			Application.Current?.MainPage ?? throw new NullReferenceException("MainPage is null."));
+
+	/// <summary>
+	/// Creates a new instance of <see cref="PopupService"/>.
+	/// </summary>
+	/// <param name="serviceProvider">The <see cref="IServiceProvider"/> implementation.</param>
+	public PopupService(IServiceProvider serviceProvider)
+	{
+		this.serviceProvider = serviceProvider;
+	}
+
+	internal static void AddTransientPopup<TPopupView, TPopupViewModel>(IServiceCollection services)
+		where TPopupView : Popup
+		where TPopupViewModel : INotifyPropertyChanged
+	{
+		viewModelToViewMappings.Add(typeof(TPopupViewModel), typeof(TPopupView));
+
+		services.AddTransient(typeof(TPopupView));
+		services.AddTransient(typeof(TPopupViewModel));
+	}
+
+	/// <inheritdoc cref="IPopupService.ShowPopup{TViewModel}()"/>
+	public void ShowPopup<TViewModel>() where TViewModel : INotifyPropertyChanged =>
+		ShowPopup(GetViewModel<TViewModel>());
+
+	/// <inheritdoc cref="IPopupService.ShowPopup{TViewModel}(TViewModel)"/>
+	public void ShowPopup<TViewModel>(TViewModel viewModel) where TViewModel : INotifyPropertyChanged
+	{
+		ArgumentNullException.ThrowIfNull(viewModel);
+
+		var popup = GetPopup(typeof(TViewModel));
+
+		CurrentPage.ShowPopup(popup);
+	}
+
+	/// <inheritdoc cref="IPopupService.ShowPopup{TViewModel}(Action{TViewModel})"/>
+	public void ShowPopup<TViewModel>(Action<TViewModel> onPresenting) where TViewModel : INotifyPropertyChanged
+	{
+		ArgumentNullException.ThrowIfNull(onPresenting);
+
+		var popup = GetPopup(typeof(TViewModel));
+
+		var viewModel = AssignBindingContext<TViewModel>(popup);
+
+		onPresenting.Invoke(viewModel);
+
+		CurrentPage.ShowPopup(popup);
+	}
+
+	/// <inheritdoc cref="IPopupService.ShowPopup{TViewModel}(Func{TViewModel, Task})"/>
+	public async void ShowPopup<TViewModel>(Func<TViewModel, Task> onPresenting) where TViewModel : INotifyPropertyChanged
```

Comment:
I think we don't need this overload, the way _sync  over async_ works, the users can have the same behavior using the method at line `53`, they just to make the `Action` async.

Reply:
I guess I wasn't sure if we wanted to allow users to wait until their async calls had finished before showing the popup rather than after. I am all for a smaller API surface so happy to remove the `Func<TViewModel, Task>` option though

Reply:
With the `Task ShowPopupAsync (Func<TViewModel, Task> onPresenting)` we will await until the Func completes to show the popup because we await it and after that, we show the popup.

Reply:
I have removed the Task options for now

Reply:
@bijington 
Just to picture the behavior on my mind... What should be the behavior when I do:

```csharp

var result = await PopupService.ShowPopup( async (vm) => {

var data = await myService.LoadDataAsync();
Items.AddRange(data);
});
```

Here the Popup should be visible during the `await myService.LoadDataAsync();` or it should wait until the `Func` finish his execution to show up?

Reply:
I expect the Popup to be visible during the `await myService.LoadDataAsync();` I have added a sample that shows this. It shows the Popup and then simulates the running of updates with the Popup showing the updates. I think this is enough, if we wanted to prevent the Popup from showing then the `Func<Task>` might need to return however I don't know if that would result in a good user experience.

The final piece of the puzzle is a clean way to allow the view model to return the value back from the Popup